modules/messages/server/services/message-to-influx.server.service.js

```diff
@@ -0,0 +1,186 @@
+
+  async.waterfall([
+    function readFirstMessage(done) {
+      // find the oldest message of the thread
+      return Message.findOne({
+        $or: [
+          {
+            userTo: userTo,
+            userFrom: userFrom
+          },
+          {
+            userTo: userFrom,
+            userFrom: userTo
+          }
+        ]
+      })
+      .sort({ created: 1 })
+      .exec(done);
+    },
+
+    function readFirstReply(firstMessage, done) {
+      // if no message was found, throw error (there is always the first message
+      // already (at least the one just saved))
+      if (!firstMessage) {
+        var err = new Error('first message not found, but should have been already saved');
+        done(err);
+      }
+
+      // is the new message the first message of the thread?
+      isFirstMessage = !!(String(firstMessage._id) === String(message._id));
+
+      // can the message be the actual first reply?
+      // - is it not the firstMessage?
+      // is the sender and receiver in different order than in the firstMessage?
+      var canBeTheFirstReply = !isFirstMessage && String(firstMessage.userTo) === String(userFrom);
+      // if this can be the oldest reply, find the oldest reply of the thread
+      if (canBeTheFirstReply) {
+        return Message.findOne({
+          userTo: firstMessage.userFrom,
+          userFrom: firstMessage.userTo
+        })
+        .sort({ created: 1 })
+        .exec(function (err, firstReply) {
+          return done(err, firstMessage, firstReply);
+        });
+      } else {
+        return done(null, firstMessage, null);
+      }
+    },
+
+    function prepareData(firstMessage, firstReply, done) {
+      // is the new message the first reply of the thread?
+      isFirstReply = !!(firstReply && String(firstReply._id) === String(message._id));
+
+      // count the reply time for statistics (miliseconds)
+      var replyTime;
+      if (isFirstReply) {
+        replyTime = firstReply.created.getTime() - firstMessage.created.getTime();
+      }
+
+      // count length of the message
+      // excluding html tags and multiple whitespace characters
+      var msgLen = textProcessor.plainText(message.content, true).length;
+
+
+      // message position in the thread
+      var position;
+      if (isFirstMessage) {
+        position = 'first';
+      } else if (isFirstReply) {
+        position = 'first_reply';
+      } else {
+        position = 'other';
+      }
+
+      var msgLenType = msgLen < config.longMessageMinimumLength ? 'short' : 'long';
+
+      // values for influxdb, using camelCase for tag and field keys
+      var fields = {
+        messageId: String(message._id), // id of message, (discussed keeping for now)
+        idFrom: String(userFrom), // id of sender
+        idTo: String(userTo), // id of receiver
+        msgLength: msgLen, // length of the content
+        replyTime: isFirstReply ? replyTime : -1, // reply time when first response only
```

Comment:
the replyTime is the time which it takes from the first message in thread till the first reply (it's not necessarily the time between first and second message). Therefore it makes sense to keep this value only with the first_reply messages. Everywhere else it is -1 (a random nonsensical value). Well, actually, maybe it shouldn't be present in the other messages at all...